docs/developer_overview.md

```diff
@@ -0,0 +1,511 @@
+# CUB Developer Overview
+
+
+This living document serves as a guide to the design of the internal structure of CUB. 
+
+The algorithms CUB provides are layered in a way that corresponds to the hierarchy of threads in CUDA where higher-level layers build on top of lower. 
+
+These layers are:
+- Thread 
+- Warp
+- Block
+- Device
+
+For example, consider `reduce`. CUB has four flavors of `reduce` that correspond to these different layers:
+
```

Comment:
This may be walking the line between developer and user guide. We can evolve this doc over time once we actually have both a user and dev guide :slightly_smiling_face:. 

```suggestion
For example, consider `reduce`. CUB has four flavors of `reduce` that correspond to these different layers. Each layer has unique semantics with respect to how they are invoked, how many threads participate, and on which thread(s) the result is valid. 

`cub::ThreadReduce` 
   - A normal function invoked and executed sequentially by a single thread that returns a valid result on that thread

`cub::WarpReduce` and `cub::BlockReduce`
   - A "cooperative" function where threads concurrently invoke the same function to execute parallel work
   - The function's return value is well-defined only on the "first" thread (lowest thread index)

`cub::DeviceReduce` 
   - A normal function invoked by a single thread that spawns additional threads to execute parallel work
   - Result is stored in the pointer provided to the function
   - Function returns a `cudaError_t` error code


The table below provides a summary of these functions:
```